src/sync_session.cpp

```diff
@@ -421,31 +448,54 @@ void SyncSession::unregister(std::unique_lock<std::mutex>& lock)
 void SyncSession::wait_for_upload_completion(std::function<void()> callback)
 {
     REALM_ASSERT(shared_from_this());
-    auto thread = std::thread([this, callback=std::move(callback), self=shared_from_this()]() {
-        {
-            std::unique_lock<std::mutex> lock(m_state_mutex);
-            if (m_session) {
-                m_session->wait_for_upload_complete_or_client_stopped();
+    auto thread = std::thread([this, callback=std::move(callback)]() {
+        std::unique_lock<std::mutex> lock(m_state_mutex);
+        if (m_state != &State::active && m_state != &State::dying) {
+            lock.unlock();
+            callback();
+        } else {
+            ++m_pending_upload_threads;
```

Comment:
This needs to be done before creating the thread to ensure that `this` still exists by the time the thread runs.